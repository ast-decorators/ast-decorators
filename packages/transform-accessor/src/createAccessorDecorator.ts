import type {ASTSimpleDecorator} from '@ast-decorators/utils/lib/common';
import type {NodePath} from '@babel/traverse';
import {
  FunctionDeclaration,
  Identifier,
  isProperty,
  PrivateName,
  VariableDeclaration,
} from '@babel/types';
import {
  AccessorInterceptorNode,
  AccessorMethodCreator,
  assert,
  createStorage,
  TransformAccessorOptions,
} from './utils';

export const createAccessorDecorator = (
  decorator: string,
  interceptor: NodePath<AccessorInterceptorNode> | undefined,
  impl: AccessorMethodCreator,
): ASTSimpleDecorator<TransformAccessorOptions> => (
  {klass, member},
  {privacy, useClassNameForStatic}: TransformAccessorOptions = {},
): void => {
  assert(decorator, member!.node, [interceptor?.node]);

  const storage = isProperty(member!.node)
    ? createStorage(klass, member!.node, privacy)
    : undefined;

  const result = impl(
    klass,
    member!,
    interceptor,
    storage?.key as Identifier | PrivateName | undefined,
    {
      preservingDecorators: member!.node.decorators,
      // @ts-expect-error: "static" is not listed in d.ts
      useClassName: !!member.node.static && !!useClassNameForStatic,
    },
  );

  const [method, declarations] = result;

  if (declarations) {
    klass.insertBefore(
      declarations as Array<FunctionDeclaration | VariableDeclaration>,
    );
  }

  if (storage) {
    member!.insertBefore(storage);
  }

  if (method) {
    member!.replaceWith(method);
  }
};
