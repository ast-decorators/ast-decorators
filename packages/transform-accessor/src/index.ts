import replaceDecorator from '@ast-decorators/utils/lib/replaceDecorator';
import type {AccessorInterceptor} from './utils';

export type AccessorDecorator = (
  get?: AccessorInterceptor,
  set?: AccessorInterceptor,
) => PropertyDecorator;

export type GetterDecorator = (get?: AccessorInterceptor) => PropertyDecorator;
export type SetterDecorator = (set?: AccessorInterceptor) => PropertyDecorator;

const transformerName = '@ast-decorators/transform-accessor';

export const accessor = replaceDecorator(
  'accessor',
  transformerName,
) as AccessorDecorator;

export const getter = replaceDecorator(
  'getter',
  transformerName,
) as GetterDecorator;

export const setter = replaceDecorator(
  'setter',
  transformerName,
) as SetterDecorator;
