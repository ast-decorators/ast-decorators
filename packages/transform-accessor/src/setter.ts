import {ASTClassMemberCallableDecorator} from '@ast-decorators/utils/lib/common';
import {NodePath} from '@babel/core';
import {
  assignmentExpression,
  blockStatement,
  CallExpression,
  ClassBody,
  classMethod,
  classPrivateMethod,
  Decorator,
  expressionStatement,
  FunctionDeclaration,
  Identifier,
  isClassPrivateProperty,
  memberExpression,
  NumericLiteral,
  PrivateName,
  StringLiteral,
  VariableDeclaration,
} from '@babel/types';
import {
  AccessorInterceptorNode,
  AccessorMethodCreator,
  createAccessorDecorator,
  ownerNode,
  prepareInterceptor,
} from './utils';

export const setter: AccessorMethodCreator = (
  klass,
  member,
  interceptor,
  storageProperty,
  {preservingDecorators, useClassName, useContext},
) => {
  const classBody = klass.get('body') as NodePath<ClassBody>;
  const valueId = classBody.scope.generateUidIdentifier('value');

  let statement: CallExpression | Identifier;
  let declarations: Array<FunctionDeclaration | VariableDeclaration>;

  if (interceptor) {
    [statement, declarations] = prepareInterceptor(
      klass,
      interceptor.node,
      valueId,
      'set',
      useContext,
      useClassName,
    );
  } else {
    statement = valueId;
    declarations = [];
  }

  const body = blockStatement([
    expressionStatement(
      assignmentExpression(
        '=',
        memberExpression(ownerNode(klass, useClassName), storageProperty),
        statement,
      ),
    ),
  ]);

  // @ts-ignore
  const {computed, key, static: _static} = member.node;

  const method = isClassPrivateProperty(member)
    ? classPrivateMethod('set', key as PrivateName, [valueId], body, _static)
    : classMethod(
        'set',
        key as Identifier | StringLiteral | NumericLiteral,
        [valueId],
        body,
        computed,
        _static,
      );

  method.decorators = preservingDecorators as Decorator[];

  return [method, declarations];
};

export const setterTransformer: ASTClassMemberCallableDecorator = (
  set?: NodePath<AccessorInterceptorNode>,
) => createAccessorDecorator('setter', set, setter);
