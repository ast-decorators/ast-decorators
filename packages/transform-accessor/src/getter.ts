import type {ASTCallableDecorator} from '@ast-decorators/utils/lib/common';
import type {NodePath} from '@babel/traverse';
import {
  assignmentExpression,
  BlockStatement,
  blockStatement,
  callExpression,
  classMethod,
  classPrivateMethod,
  Decorator,
  ExpressionStatement,
  expressionStatement,
  FunctionDeclaration,
  identifier,
  Identifier,
  isClassPrivateProperty,
  isIdentifier,
  isMethod,
  memberExpression,
  NumericLiteral,
  PrivateName,
  ReturnStatement,
  returnStatement,
  StringLiteral,
  thisExpression,
  variableDeclaration,
  VariableDeclaration,
  variableDeclarator,
} from '@babel/types';
import {createAccessorDecorator} from './createAccessorDecorator';
import {
  AccessorInterceptorNode,
  AccessorMethodCreator,
  ownerNode,
  prepareInterceptor,
  TransformAccessorOptions,
} from './utils';

const resultKey = 'binding:result-var';

const prepareResult = (
  body: NodePath<BlockStatement>,
): [Identifier, (VariableDeclaration | ExpressionStatement)?] => {
  const {argument} = body.node.body[
    body.node.body.length - 1
  ] as ReturnStatement;

  if (isIdentifier(argument)) {
    return [argument];
  }

  const member = body.parentPath;
  const data: Identifier | undefined = member.getData(resultKey);

  if (data) {
    return [
      data,
      expressionStatement(
        assignmentExpression('=', data, argument ?? identifier('undefined')),
      ),
    ];
  }

  const resultId = body.scope.generateUidIdentifier('result');
  const resultDeclaration = variableDeclaration('let', [
    variableDeclarator(resultId, argument),
  ]);

  member.setData(resultKey, resultId);

  return [resultId, resultDeclaration];
};

// For properties, getter decorator creates a get method with a call to get
// interceptor function.
// For methods, it puts a call to the interceptor function between the existing
// code and the return statement.
//
// Original:
//
//   @getter(get)
//   get foo() {
//     return SOMETHING;
//   }
//
// Transformed
//
//   get foo() {
//     const _result = get(SOMETHING, this); // interceptor function
//
//     return _result;
//   }
export const getter: AccessorMethodCreator = (
  klass,
  member,
  interceptor,
  storageProperty,
  {preservingDecorators, useClassName},
) => {
  const declarations: Array<FunctionDeclaration | VariableDeclaration> = [];

  const [interceptorId, interceptorDeclaration] = interceptor
    ? prepareInterceptor(klass, interceptor.node, 'get')
    : [];

  if (interceptorDeclaration) {
    declarations.push(interceptorDeclaration);
  }

  let newBody: BlockStatement;

  if (isMethod(member.node)) {
    if (!interceptorId) {
      return null;
    }

    const body = member.get('body') as NodePath<BlockStatement>;

    const originalBody = body.node.body.slice(0, -1);

    const [resultId, resultDeclaration] = prepareResult(body);

    newBody = blockStatement([
      ...originalBody,
      ...(resultDeclaration ? [resultDeclaration] : []),
      returnStatement(
        callExpression(interceptorId, [resultId, thisExpression()]),
      ),
    ]);
  } else {
    const property = memberExpression(
      ownerNode(klass.node, useClassName),
      storageProperty!,
    );

    const result = interceptorId
      ? callExpression(interceptorId, [property, thisExpression()])
      : property;

    newBody = blockStatement([returnStatement(result)]);
  }

  // @ts-expect-error: "computed" do not exist on the ClassMemberProperty (it
  // will simply be undefined) and "static" is not listed in d.ts
  const {computed, key, static: _static} = member.node;

  const method = isClassPrivateProperty(member)
    ? classPrivateMethod('get', key as PrivateName, [], newBody, _static)
    : classMethod(
        'get',
        key as Identifier | StringLiteral | NumericLiteral,
        [],
        newBody,
        computed,
        _static,
      );

  method.decorators = preservingDecorators as Decorator[];

  return [method, declarations];
};

export const getterTransformer: ASTCallableDecorator<
  [NodePath<AccessorInterceptorNode>?],
  TransformAccessorOptions
> = get => createAccessorDecorator('getter', get, getter);
