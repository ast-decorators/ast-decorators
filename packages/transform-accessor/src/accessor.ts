import checkSuitability from '@ast-decorators/utils/lib/checkSuitability';
import {
  ASTClassMemberDecorator,
  ClassMemberProperty,
} from '@ast-decorators/utils/lib/common';
import {DecoratorMetadata} from '@ast-decorators/utils/lib/metadata';
import {NodePath} from '@babel/core';
import {Decorator, Identifier, PrivateName} from '@babel/types';
import {getter} from './getter';
import {setter} from './setter';
import shouldInterceptorUseContext from '@ast-decorators/utils/lib/shouldInterceptorUseContext';
import {
  AccessorInterceptorNode,
  assert,
  createStorage,
  TransformAccessorOptions,
} from './utils';

export const accessorTransformer = (
  get?: NodePath<AccessorInterceptorNode>,
  set?: NodePath<AccessorInterceptorNode>,
): ASTClassMemberDecorator<TransformAccessorOptions> => (
  klass,
  member: NodePath<ClassMemberProperty>,
  {
    interceptorContext,
    privacy,
    singleAccessorDecorators,
    useClassNameForStatic,
  }: TransformAccessorOptions = {},
  {filename},
) => {
  assert('accessor', member, [get, set]);

  const decorators = member.node.decorators
    ? (member.get('decorators') as ReadonlyArray<NodePath<Decorator>>)
    : null;

  // @ts-ignore
  const useClassName = !!member.node.static && !!useClassNameForStatic;

  const storage = createStorage(klass, member, privacy);
  const [getMethod, getterDeclarations] = getter(
    klass,
    member,
    get,
    storage.key as Identifier | PrivateName,
    {
      preservingDecorators: decorators?.map(({node}) => node) ?? null,
      useClassName,
      useContext: shouldInterceptorUseContext(
        get,
        interceptorContext,
        filename,
      ),
    },
  );

  const bothAccessorsDecorators = decorators?.filter(decorator => {
    const {importSource, originalImportName} = new DecoratorMetadata(decorator);

    return !checkSuitability(
      {
        name: originalImportName,
        source: importSource?.node.value,
      },
      singleAccessorDecorators,
      filename,
    );
  });

  const [setMethod, setterDeclarations] = setter(
    klass,
    member,
    set,
    storage.key as Identifier | PrivateName,
    {
      preservingDecorators:
        bothAccessorsDecorators?.map(({node}) => node) ?? null,
      useClassName,
      useContext: shouldInterceptorUseContext(
        set,
        interceptorContext,
        filename,
      ),
    },
  );

  klass.insertBefore([...getterDeclarations, ...setterDeclarations]);
  member.replaceWithMultiple([storage, getMethod, setMethod]);
};
