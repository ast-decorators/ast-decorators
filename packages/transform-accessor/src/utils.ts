import ASTDecoratorsError from '@ast-decorators/utils/lib/ASTDecoratorsError';
import type {SuitabilityFactors} from '@ast-decorators/utils/lib/checkSuitability';
import type {
  ClassMemberMethod,
  ClassMemberProperty,
  PrivacyType,
} from '@ast-decorators/utils/lib/common';
import createPropertyByPrivacy from '@ast-decorators/utils/lib/createPropertyByPrivacy';
import getMemberName from '@ast-decorators/utils/lib/getMemberName';
import {ClassMember} from '@ast-decorators/utils/src/common';
import template from '@babel/template';
import type {NodePath, Scope} from '@babel/traverse';
import {
  ArrayPattern,
  ArrowFunctionExpression,
  Class,
  cloneNode,
  Decorator,
  FunctionDeclaration,
  functionDeclaration,
  FunctionExpression,
  Identifier,
  isArrayPattern,
  isArrowFunctionExpression,
  isClassMethod,
  isClassPrivateMethod,
  isClassPrivateProperty,
  isClassProperty,
  isFunctionExpression,
  isIdentifier,
  isMemberExpression,
  isObjectPattern,
  MemberExpression,
  ObjectPattern,
  PrivateName,
  ThisExpression,
  thisExpression,
  variableDeclaration,
  VariableDeclaration,
  variableDeclarator,
} from '@babel/types';

export type TransformedNode = readonly [
  ClassMemberMethod,
  ReadonlyArray<FunctionDeclaration | VariableDeclaration>,
];

export type TransformAccessorOptions = Readonly<{
  privacy?: PrivacyType;
  singleAccessorDecorators?: SuitabilityFactors;
  transformerPath?: string;
  useClassNameForStatic?: boolean;
}>;

export type AccessorInterceptor = (value: any) => any;
export type AccessorInterceptorNode =
  | FunctionExpression
  | ArrowFunctionExpression
  | Identifier
  | MemberExpression;

export type AccessorMethodCreatorOptions = Readonly<{
  useClassName: boolean;
  preservingDecorators: Decorator[] | null;
}>;

export type AccessorMethodCreator = (
  klass: NodePath<Class>,
  member: NodePath<ClassMember>,
  interceptor: NodePath<AccessorInterceptorNode> | undefined,
  storage: Identifier | PrivateName | undefined,
  options: AccessorMethodCreatorOptions,
) => TransformedNode | null;

export const isGetter = (member: ClassMemberMethod): boolean =>
  member.kind === 'get';
export const isSetter = (member: ClassMemberMethod): boolean =>
  member.kind === 'set';

export const assert = (
  decorator: string,
  member: ClassMember,
  interceptors: ReadonlyArray<AccessorInterceptorNode | undefined>,
): void => {
  if (
    !(isClassProperty(member) || isClassPrivateProperty(member)) &&
    !(
      (isClassMethod(member) || isClassPrivateMethod(member)) &&
      (isGetter(member) || isSetter(member))
    )
  ) {
    throw new ASTDecoratorsError(
      `Applying @${decorator} decorator to something other than property or accessor is not allowed`,
    );
  }

  for (const interceptor of interceptors) {
    if (
      interceptor &&
      !isFunctionExpression(interceptor) &&
      !isArrowFunctionExpression(interceptor) &&
      !isIdentifier(interceptor) &&
      !isMemberExpression(interceptor)
    ) {
      throw new ASTDecoratorsError(
        'Accessor interceptor can only be function, free variable or object property',
      );
    }
  }
};

export const createStorage = (
  klass: NodePath<Class>,
  member: ClassMemberProperty,
  privacy: PrivacyType = 'hard',
): ClassMemberProperty =>
  createPropertyByPrivacy(privacy, getMemberName(member), klass, {
    // @ts-expect-error: "static" is not listed in d.ts
    static: member.static,
    value: member.value,
  });

const declarator = template(`const VAR = FUNCTION`);

export const ownerNode = (
  klass: Class,
  useClassName: boolean,
): Identifier | ThisExpression =>
  useClassName && klass.id ? cloneNode(klass.id) : thisExpression();

export const prepareInterceptor = (
  klass: NodePath<Class>,
  interceptor: AccessorInterceptorNode,
  type: 'get' | 'set',
): readonly [
  Identifier | MemberExpression,
  FunctionDeclaration | VariableDeclaration | null,
] => {
  const isArrow = isArrowFunctionExpression(interceptor);
  const isRegular = isFunctionExpression(interceptor);

  const interceptorId =
    isArrow || isRegular
      ? klass.parentPath.scope.generateUidIdentifier(type)
      : (interceptor as Identifier | MemberExpression);

  return [
    interceptorId,
    isArrow
      ? (declarator({
          FUNCTION: interceptor,
          VAR: interceptorId as Identifier,
        }) as VariableDeclaration)
      : isRegular
      ? functionDeclaration(
          interceptorId as Identifier,
          (interceptor as FunctionExpression).params,
          (interceptor as FunctionExpression).body,
        )
      : null,
  ];
};

export const unifyValueParameter = (
  scope: Scope,
  rawValue: Identifier | ArrayPattern | ObjectPattern,
): [Identifier, VariableDeclaration?] => {
  if (isObjectPattern(rawValue) || isArrayPattern(rawValue)) {
    const value = scope.generateUidIdentifier('value');

    return [
      value,
      variableDeclaration('const', [variableDeclarator(rawValue, value)]),
    ];
  }

  return [rawValue];
};
