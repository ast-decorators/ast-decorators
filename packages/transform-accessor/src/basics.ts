import {
  classMethod,
  classPrivateMethod,
  cloneNode,
} from '@ast-decorators/utils/lib/babelFixes';
import {ClassMember, ClassMemberMethod} from '@ast-decorators/utils/src/common';
import {NodePath} from '@babel/traverse';
import {
  assignmentExpression,
  blockStatement,
  Class,
  expressionStatement,
  identifier,
  Identifier,
  isMethod,
  isPrivate,
  memberExpression,
  NumericLiteral,
  PrivateName,
  returnStatement,
  StringLiteral,
} from '@babel/types';
import {AccessorMethodCreatorOptions, ownerNode} from './utils';

export const basicGetter = (
  klass: NodePath<Class>,
  member: NodePath<ClassMember>,
  storageProperty: Identifier | PrivateName | undefined,
  {preservingDecorators, useClassName}: AccessorMethodCreatorOptions,
): ClassMemberMethod => {
  if (isMethod(member.node)) {
    const method = cloneNode(member.node);
    method.decorators = preservingDecorators;

    return method;
  }

  const body = blockStatement([
    returnStatement(
      memberExpression(ownerNode(klass.node, useClassName), storageProperty!),
    ),
  ]);

  // @ts-expect-error: "computed" do not exist on the ClassPrivateProperty (it
  // will simply be undefined) and "static" is not listed in d.ts
  const {computed, key, static: _static} = member.node;

  return isPrivate(member)
    ? classPrivateMethod(
        'get',
        key as PrivateName,
        [],
        body,
        preservingDecorators,
        _static,
      )
    : classMethod(
        'get',
        key as Identifier | StringLiteral | NumericLiteral,
        [],
        body,
        preservingDecorators,
        computed,
        _static,
      );
};

export const basicSetter = (
  klass: NodePath<Class>,
  member: NodePath<ClassMember>,
  storageProperty: Identifier | PrivateName | undefined,
  {preservingDecorators, useClassName}: AccessorMethodCreatorOptions,
): ClassMemberMethod => {
  if (isMethod(member.node)) {
    const method = cloneNode(member.node);
    method.decorators = preservingDecorators;

    return method;
  }

  const valueId = identifier('value');
  const params = [valueId];

  const property = memberExpression(
    ownerNode(klass.node, useClassName),
    storageProperty!,
  );

  const body = blockStatement([
    expressionStatement(assignmentExpression('=', property, valueId)),
  ]);

  // @ts-expect-error: "computed" do not exist on the ClassPrivateProperty (it
  // will simply be undefined) and "static" is not listed in d.ts
  const {computed, key, static: _static} = member.node;

  return isPrivate(member)
    ? classPrivateMethod(
        'set',
        key as PrivateName,
        params,
        body,
        preservingDecorators,
        _static,
      )
    : classMethod(
        'set',
        key as Identifier | StringLiteral | NumericLiteral,
        params,
        body,
        preservingDecorators,
        computed,
        _static,
      );
};
