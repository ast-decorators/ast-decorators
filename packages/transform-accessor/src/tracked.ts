import {cloneNode} from '@ast-decorators/utils/lib/babelFixes';
import type {
  ASTCallableDecorator,
  ClassMember,
  ClassMemberProperty,
} from '@ast-decorators/utils/lib/common';
import hoistFunctionParameter from '@ast-decorators/utils/lib/hoistFunctionParameter';
import {ClassMemberMethod} from '@ast-decorators/utils/src/common';
import type {NodePath, Scope} from '@babel/traverse';
import {
  ArrayPattern,
  blockStatement,
  callExpression,
  Class,
  Decorator,
  expressionStatement,
  Identifier,
  isMethod,
  isProperty,
  MemberExpression,
  ObjectPattern,
  PrivateName,
  thisExpression,
} from '@babel/types';
import {basicGetter, basicSetter} from './basics';
import {
  AccessorInterceptorNode,
  applyChanges,
  assert,
  createStorage,
  prepareSetterDecorators,
  TransformAccessorOptions,
  TwinAccessorOptions,
  TwinAccessorTransformedNode,
  unifyValueParameter,
} from './utils';

const addTracker = (
  method: ClassMemberMethod,
  track: Identifier | MemberExpression | undefined,
  scope: Scope,
) => {
  if (track) {
    const [rawValue] = method.params;

    const [valueId, valueSupportDeclaration] = unifyValueParameter(
      scope,
      // There is no reason to set the default value to value parameter
      rawValue as Identifier | ArrayPattern | ObjectPattern,
    );

    method.body = blockStatement([
      ...(valueSupportDeclaration ? [valueSupportDeclaration] : []),
      ...method.body.body,
      expressionStatement(callExpression(track, [thisExpression(), valueId])),
    ]);
  }

  return method;
};

export const tracked = (
  klass: NodePath<Class>,
  member: NodePath<ClassMember>,
  interceptor: NodePath<AccessorInterceptorNode> | undefined,
  {
    privacy,
    useClassNameForStatic,
    ...decoratorsPreparationOptions
  }: TwinAccessorOptions,
): TwinAccessorTransformedNode => {
  const isMemberProperty = isProperty(member.node);

  const decorators = member.node.decorators
    ? (member.get('decorators') as ReadonlyArray<NodePath<Decorator>>)
    : null;

  const useClassName = 'static' in member.node && !!useClassNameForStatic;

  const storage = isMemberProperty
    ? createStorage(klass, member.node as ClassMemberProperty, privacy)
    : undefined;

  const [trackId, trackDeclaration] = interceptor
    ? hoistFunctionParameter(interceptor.node, 'track', klass.parentPath.scope)
    : [];

  const getter = isMemberProperty
    ? basicGetter(
        klass,
        member,
        storage?.key as Identifier | PrivateName | undefined,
        {
          preservingDecorators: decorators?.map(({node}) => node) ?? null,
          useClassName,
        },
      )
    : undefined;

  const setterPreservedDecorators = decorators
    ? prepareSetterDecorators(decorators, decoratorsPreparationOptions)
        // We need to add decorator nodes to binding registry only if we create two
        // nodes (getter and setter) instead of one (property).
        .map(({node}) => (isMemberProperty ? cloneNode(node) : node))
    : null;

  const setter =
    isMemberProperty || (isMethod(member.node) && member.node.kind === 'set')
      ? addTracker(
          basicSetter(
            klass,
            member,
            storage?.key as Identifier | PrivateName | undefined,
            {
              preservingDecorators: setterPreservedDecorators,
              useClassName,
            },
          ),
          trackId,
          member.scope,
        )
      : undefined;

  return [
    [getter],
    [setter, trackDeclaration ? [trackDeclaration] : undefined],
    storage,
  ];
};

export const trackedTransformer: ASTCallableDecorator<
  [NodePath<AccessorInterceptorNode>?],
  TransformAccessorOptions
> = (interceptor) => (
  {klass, member},
  options: TransformAccessorOptions = {},
  {filename},
) => {
  assert('tracked', member!.node, [interceptor?.node]);

  applyChanges(
    klass,
    member!,
    tracked(klass, member!, interceptor, {...options, filename}),
  );
};
