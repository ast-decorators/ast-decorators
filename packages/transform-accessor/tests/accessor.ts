import {transformFile as _transformFile} from '../../../utils/testing';
import {accessor} from '../src';
import commonOptions from './fixtures/options';

const transformFile = async (
  fixture: string,
  options?: string | object,
): ReturnType<typeof _transformFile> =>
  _transformFile(__dirname, 'accessor', fixture, options);

describe('@ast-decorators/transform-accessor', () => {
  describe('@accessor', () => {
    it('compiles without interceptors', async () => {
      const {code} = await transformFile('default', commonOptions);
      expect(code).toMatchSnapshot();
    });

    it('compiles for static property', async () => {
      const {code} = await transformFile('static-property', commonOptions);
      expect(code).toMatchSnapshot();
    });

    it('preserves following decorators for both accessors', async () => {
      const {code} = await transformFile(
        'preserve-decorators-for-both',
        commonOptions,
      );
      expect(code).toMatchSnapshot();
    });

    it('preserves following decorators only for getter', async () => {
      const {code} = await transformFile('preserve-decorators-for-single');
      expect(code).toMatchSnapshot();
    });

    it('throws an error if transformer is not plugged in', () => {
      // @ts-expect-error: Here the runtime replacement used. It does not
      // require arguments
      expect(() => accessor()()).toThrowError(
        "Decorator @accessor won't work because @ast-decorators/transform-accessor/lib/transformer" +
          'is not plugged in. You have to add it to your Babel config',
      );
    });
  });
});
