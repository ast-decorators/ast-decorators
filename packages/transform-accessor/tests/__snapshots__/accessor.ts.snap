// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@ast-decorators/transform-accessor @accessor compiles for static property 1`] = `
"class Foo {
  static #_bar;

  static get bar() {
    return this.#_bar;
  }

  static set bar(value) {
    this.#_bar = value;
  }

}"
`;

exports[`@ast-decorators/transform-accessor @accessor compiles without interceptors 1`] = `
"class Foo {
  #_bar;

  get bar() {
    return this.#_bar;
  }

  set bar(value) {
    this.#_bar = value;
  }

}"
`;

exports[`@ast-decorators/transform-accessor @accessor preserves following decorators for both accessors 1`] = `
"import { other } from 'decorators';

class Foo {
  #_bar;

  @other()
  get bar() {
    return this.#_bar;
  }

  @other()
  set bar(value) {
    this.#_bar = value;
  }

}"
`;

exports[`@ast-decorators/transform-accessor @accessor preserves following decorators only for getter 1`] = `
"import { other } from 'decorators';

class Foo {
  #_bar;

  @other()
  get bar() {
    return this.#_bar;
  }

  set bar(value) {
    this.#_bar = value;
  }

}"
`;
